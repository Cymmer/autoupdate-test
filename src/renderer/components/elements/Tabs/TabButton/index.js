import React from 'react';
import cn from 'classnames';
import PropTypes from 'prop-types';
import styles from '../tabs.module.scss';

import IconButton from '../../Button/IconButton';
import iconButtonTypes from '../../Button/constants/iconButtonTypes';

import tabTypes from '../constants/tabTypes';

const TabButton = ({
  innerRef,
  className,
  children,
  type,
  active,
  onClick,
  closeAction,
  id,
  hasNewData,
}) => (
  <button
    ref={(e) => {
      if (innerRef) {
        innerRef.current = e;
      }
    }}
    className={cn(className, {
      [styles[`Tab___${type}___active`]]: active,
      [styles[`Tab___${type}`]]: !active,
      [styles[`Tab___${type}___hasNewData`]]: hasNewData,
      [styles.Tab___withClose]: closeAction && active,
    })}
    data-test="tabButton"
    id={id}
    type="button"
    onClick={onClick}
  >
    {children}
    {closeAction && active && (
      <IconButton
        className={styles.Tab_close}
        data-test="closeButton"
        icon="close"
        type={iconButtonTypes.SOLID.XS}
        onClick={closeAction}
      />
    )}
  </button>
);

TabButton.defaultProps = {
  className: null,
  active: false,
  type: tabTypes.HORIZONTAL.LG,
  closeAction: null,
  id: null,
  hasNewData: false,
};

TabButton.propTypes = {
  innerRef: PropTypes.object,
  className: PropTypes.string,
  active: PropTypes.bool,
  onClick: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
  type: PropTypes.oneOf([
    tabTypes.HORIZONTAL.LG,
    tabTypes.HORIZONTAL.SM,
    tabTypes.VERTICAL.LG,
    tabTypes.VERTICAL.SM,
  ]),
  closeAction: PropTypes.func,
  id: PropTypes.string,
  hasNewData: PropTypes.bool,
};

export default TabButton;
