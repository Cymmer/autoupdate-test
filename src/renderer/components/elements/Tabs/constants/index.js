export { default as tabTypes } from './tabTypes';
export { default as tabKinds } from './tabKinds';
