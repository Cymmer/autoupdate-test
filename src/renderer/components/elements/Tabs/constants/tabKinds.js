const tabKinds = {
  BUTTON: 'button',
  LINK: 'link',
};

export default tabKinds;
