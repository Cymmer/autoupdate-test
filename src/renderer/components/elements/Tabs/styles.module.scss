@use '../../../scss' as *;

@mixin shine {
  width: 120px;
  margin: 24px;
  padding: 12px;
}

.Tabs {
  &_button,
  &_link {
    text-transform: capitalize;
  }

  &___horizontal___lg,
  &___horizontal___sm {
    display: flex;
    overflow: auto;

    > .Tabs_shine {
      @include shine;
    }
  }

  &___vertical___lg,
  &___vertical___sm {
    > .Tabs_shine {
      @include shine;
    }
  }

  &___vertical___lg {
    display: grid;
    grid-gap: 12px;
  }

  &___vertical___sm {
    display: flex;
    flex-direction: column;

    > .Tabs_button,
    > .Tabs_link {
      margin: 0 0 0.25em 0;
      white-space: nowrap;

      @include breakpoint(sm) {
        margin: 0 0.5em 0 0;
      }
    }

    @include breakpoint(sm) {
      flex-direction: row;
    }
  }

  &___loading {
    padding: 16px;
  }
}
