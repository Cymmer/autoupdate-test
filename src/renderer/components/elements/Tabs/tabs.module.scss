@import '../../../scss/variables';
@import '../../../scss/mixins';
@import '../../../scss/themes';

@mixin base($font-size, $line-height, $padding) {
  font-family: Montserrat, sans-serif;
  font-size: $font-size;
  line-height: $line-height;
  font-weight: 700;
  font-style: normal;
  padding: $padding;
  background-color: transparent;
  border: none;
  cursor: pointer;
  width: max-content;
  text-decoration: none;
  text-align: center;
  @include ease(0.2s);
  position: relative;

  @content;
}

@mixin horizontal-lg($state) {
  @include base(1rem, 1.25, 16px 32px) {
    @if $state == unselected {
      @include theme {
        color: t('tabTextColor');
      }
      border-bottom: 2px solid transparent;
    } @else if $state == selected {
      color: $blue-300;
      border-bottom: 2px solid $blue-300;
    }

    &:hover {
      background-color: transparentize($blue-300, 0.9);
    }

    &:active {
      background-color: transparentize($blue-300, 0.8);
    }
  }
}

@mixin horizontal-sm($state) {
  @include base(0.75rem, 1.67, 8px 16px 4px 16px) {
    @if $state == unselected {
      @include theme {
        color: t('tabTextColor');
      }
      border-bottom: 2px solid transparent;
    } @else if $state == selected {
      @include theme {
        color: t('tabTextColor--active');
      }
      border-bottom: 2px solid $blue-300;
    }

    &:hover {
      color: $neutral-700;
    }
  }
}

@mixin vertical($font-size, $line-height, $padding, $state) {
  @include base($font-size, $line-height, $padding) {
    border-radius: 100px;
    margin: 2px 0;

    @if $state == unselected {
      @include theme {
        color: t('tabTextColor');
      }
    } @else if $state == selected {
      color: $blue-300;
      background-color: transparentize($blue-300, 0.9);
    }

    &:hover {
      background-color: transparentize($blue-300, 0.9);
    }

    &:active {
      background-color: transparentize($blue-300, 0.8);
    }
  }
}

@mixin vertical-lg($state) {
  @include vertical(1rem, 1.25, 16px, $state);
}

@mixin vertical-sm($state) {
  @include vertical(0.875rem, 1.67, 12px 16px, $state);
}

.Tab {
  &___horizontal___lg {
    @include horizontal-lg(unselected);

    &___active {
      @include horizontal-lg(selected);
    }

    &___hasNewData {
      &:after {
        content: '';
        width: 6px;
        height: 6px;
        background-color: $red-300;
        border-radius: 100%;
        border: 2px solid transparent;
        position: absolute;
        top: 14px;
        right: 20px;
      }
    }
  }

  &___locked {
    cursor: not-allowed;

    &:hover {
      background-color: transparent !important;
    }
  }

  &___horizontal___sm {
    @include horizontal-sm(unselected);

    &___active {
      @include horizontal-sm(selected);
    }

    &___hasNewData {
      &:after {
        content: '';
        width: 10px;
        height: 10px;
        background-color: $red-300;
        border-radius: 100%;
        border: 2px solid;
        @include theme {
          border-color: t('cardBackgroundColor');
        }
        position: absolute;
        top: 6px;
        right: 6px;
      }
    }
  }

  &___vertical___lg {
    @include vertical-lg(unselected);

    &___active {
      @include vertical-lg(selected);
    }
  }

  &___vertical___sm {
    @include vertical-sm(unselected);

    &___active {
      @include vertical-sm(selected);
    }
  }

  &___withClose {
    display: flex;
    align-items: center;
    padding: 8px 4px 4px 16px;
    transition: none;
  }

  &_close {
    margin: 0 0 0 4px;
  }

  &_linkWrapper {
    display: flex;
    position: relative;
  }

  &_lockedIcon {
    position: absolute;
    top: 17px;
    left: 12px;
    color: $yellow-400;
    z-index: 1;
  }
}
