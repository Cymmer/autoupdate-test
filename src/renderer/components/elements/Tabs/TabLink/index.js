import React from 'react';
import { Link } from 'react-router-dom';
import cn from 'classnames';
import GLOBALS from 'codechum-app-globals';
import PropTypes from 'prop-types';

import { Icon, Tooltip } from 'components/elements';
import { STUDENT_ROUTES } from 'screen-wrappers/Student/constants';
import styles from '../tabs.module.scss';

import IconButton from '../../Button/IconButton';
import iconButtonTypes from '../../Button/constants/iconButtonTypes';
import tabTypes from '../constants/tabTypes';

const handleClickIfLocked = (e) => {
  e.preventDefault();
};

const TabLink = ({
  className,
  children,
  type,
  active,
  to,
  closeAction,
  id,
  hasNewData,
  isLocked,
}) => {
  const link = (
    <Link
      className={cn(className, {
        [styles[`Tab___${type}___active`]]: active,
        [styles[`Tab___${type}`]]: !active,
        [styles[`Tab___${type}___hasNewData`]]: hasNewData,
        [styles.Tab___withClose]: closeAction && active,
        [styles.Tab___locked]: isLocked,
      })}
      data-test="tabLink"
      id={id}
      to={to}
      onClick={isLocked ? handleClickIfLocked : null}
    >
      {children}
      {closeAction && active && (
        <IconButton
          className={styles.Tab_close}
          data-test="closeButton"
          icon="close"
          type={iconButtonTypes.SOLID.XS}
          onClick={closeAction}
        />
      )}
    </Link>
  );

  return !isLocked ? (
    link
  ) : (
    <span className={styles.Tab_linkWrapper}>
      <Tooltip content={GLOBALS.MESSAGES.PREMIUM_FEATURE_LOCK_MESSAGE}>
        <a href={STUDENT_ROUTES.SHOP} rel="noreferrer" target="_blank">
          <Icon className={styles.Tab_lockedIcon} icon="lock" />
        </a>
      </Tooltip>
      {link}
    </span>
  );
};

TabLink.defaultProps = {
  className: null,
  active: false,
  type: tabTypes.HORIZONTAL.LG,
  to: '#',
  closeAction: null,
  id: null,
  hasNewData: false,
  isLocked: false,
};

TabLink.propTypes = {
  className: PropTypes.string,
  active: PropTypes.bool,
  to: PropTypes.string,
  children: PropTypes.node.isRequired,
  type: PropTypes.oneOf([
    tabTypes.HORIZONTAL.LG,
    tabTypes.HORIZONTAL.SM,
    tabTypes.VERTICAL.LG,
    tabTypes.VERTICAL.SM,
  ]),
  closeAction: PropTypes.func,
  id: PropTypes.string,
  hasNewData: PropTypes.bool,
  isLocked: PropTypes.bool,
};

export default TabLink;
