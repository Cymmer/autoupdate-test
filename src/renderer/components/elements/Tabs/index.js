import React from 'react';
import cn from 'classnames';
import PropTypes from 'prop-types';
import styles from './styles.module.scss';

import Shine from '../Shine';

import tabTypes from './constants/tabTypes';
import tabKinds from './constants/tabKinds';

import TabButton from './TabButton';
import TabLink from './TabLink';

const Tabs = ({
  type,
  activeTab,
  tabs,
  className,
  tabClassName,
  isLoading,
}) => (
  <div
    className={cn(className, styles[`Tabs___${type}`], {
      [styles.Tabs___loading]: isLoading,
    })}
  >
    {isLoading ? (
      <>
        <Shine className={styles.Tabs_shine} />
        <Shine className={styles.Tabs_shine} />
      </>
    ) : (
      tabs?.map(
        ({
          ref,
          name,
          value,
          kind,
          action,
          closeAction,
          id,
          hasNewData,
          isLocked,
        }) =>
          kind === tabKinds.BUTTON ? (
            <TabButton
              key={value}
              active={activeTab === value}
              className={cn(styles.Tabs_button, tabClassName)}
              closeAction={
                type === tabTypes.HORIZONTAL.SM && closeAction
                  ? closeAction
                  : null
              }
              hasNewData={hasNewData}
              id={id}
              innerRef={ref}
              type={type}
              onClick={action}
            >
              {name}
            </TabButton>
          ) : (
            <TabLink
              key={value}
              active={activeTab === value}
              className={cn(styles.Tabs_link, tabClassName)}
              closeAction={
                type === tabTypes.HORIZONTAL.SM && closeAction
                  ? closeAction
                  : null
              }
              hasNewData={hasNewData}
              id={id}
              isLocked={isLocked}
              to={action}
              type={type}
            >
              {name}
            </TabLink>
          )
      )
    )}
  </div>
);

Tabs.defaultProps = {
  className: null,
  tabClassName: null,
  isLoading: false,
  activeTab: null,
};

Tabs.propTypes = {
  className: PropTypes.string,
  tabClassName: PropTypes.string,
  isLoading: PropTypes.bool,
  type: PropTypes.oneOf([
    tabTypes.HORIZONTAL.LG,
    tabTypes.HORIZONTAL.SM,
    tabTypes.VERTICAL.LG,
    tabTypes.VERTICAL.SM,
  ]).isRequired,
  tabs: PropTypes.arrayOf(
    PropTypes.shape({
      ref: PropTypes.object,
      name: PropTypes.string.isRequired,
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
        .isRequired,
      kind: PropTypes.oneOf([tabKinds.BUTTON, tabKinds.LINK]).isRequired,
      // if type is BUTTON -> func, else if type is LINK -> string
      action: PropTypes.oneOfType([PropTypes.func, PropTypes.string])
        .isRequired,
      isLocked: PropTypes.bool,
      closeAction: PropTypes.func,
      id: PropTypes.string,
    })
  ).isRequired,
  activeTab: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

export default Tabs;
