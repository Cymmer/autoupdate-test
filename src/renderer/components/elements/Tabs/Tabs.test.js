import React from 'react';
import { shallow } from 'enzyme';
import Tabs from './index';
import tabTypes from './constants/tabTypes';
import tabKinds from './constants/tabKinds';

describe('RENDERING - Tabs', () => {
  it('should render with default props', () => {
    const component = shallow(<Tabs tabs={[]} type={tabTypes.HORIZONTAL.LG} />);
    expect(component).toMatchSnapshot();
  });

  it('should render the loading state if it is loading', () => {
    const component = shallow(
      <Tabs isLoading tabs={[]} type={tabTypes.HORIZONTAL.LG} />
    );
    expect(component).toMatchSnapshot();
  });

  it('should render its children appropriately if it is not loading', () => {
    const component = shallow(
      <Tabs
        tabs={[
          {
            name: 'Test 1',
            value: 'test1',
            kind: tabKinds.BUTTON,
            action: () => {},
            closeAction: () => {},
          },
          {
            name: 'Test 2',
            value: 'test2',
            kind: tabKinds.BUTTON,
            action: () => {},
          },
          {
            name: 'Test 3',
            value: 'test3',
            kind: tabKinds.LINK,
            action: 'www.codechum.com',
          },
          {
            name: 'Test 4',
            value: 'test4',
            kind: tabKinds.LINK,
            action: 'www.codechum.com/privacy-policy',
          },
          {
            name: 'Test 5',
            value: 'test5',
            kind: tabKinds.LINK,
            action: () => {},
            isLocked: true,
          },
        ]}
        type={tabTypes.HORIZONTAL.LG}
      />
    );
    expect(component).toMatchSnapshot();
  });
});
