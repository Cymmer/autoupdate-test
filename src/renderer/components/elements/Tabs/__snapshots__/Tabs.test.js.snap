// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RENDERING - Tabs should render its children appropriately if it is not loading 1`] = `
<div
  className="Tabs___horizontal___lg"
>
  <TabButton
    active={false}
    className="Tabs_button"
    closeAction={null}
    hasNewData={false}
    id={null}
    key="test1"
    onClick={[Function]}
    type="horizontal___lg"
  >
    Test 1
  </TabButton>
  <TabButton
    active={false}
    className="Tabs_button"
    closeAction={null}
    hasNewData={false}
    id={null}
    key="test2"
    onClick={[Function]}
    type="horizontal___lg"
  >
    Test 2
  </TabButton>
  <TabLink
    active={false}
    className="Tabs_link"
    closeAction={null}
    hasNewData={false}
    id={null}
    isLocked={false}
    key="test3"
    to="www.codechum.com"
    type="horizontal___lg"
  >
    Test 3
  </TabLink>
  <TabLink
    active={false}
    className="Tabs_link"
    closeAction={null}
    hasNewData={false}
    id={null}
    isLocked={false}
    key="test4"
    to="www.codechum.com/privacy-policy"
    type="horizontal___lg"
  >
    Test 4
  </TabLink>
  <TabLink
    active={false}
    className="Tabs_link"
    closeAction={null}
    hasNewData={false}
    id={null}
    isLocked={true}
    key="test5"
    to={[Function]}
    type="horizontal___lg"
  >
    Test 5
  </TabLink>
</div>
`;

exports[`RENDERING - Tabs should render the loading state if it is loading 1`] = `
<div
  className="Tabs___horizontal___lg Tabs___loading"
>
  <Shine
    className="Tabs_shine"
  />
  <Shine
    className="Tabs_shine"
  />
</div>
`;

exports[`RENDERING - Tabs should render with default props 1`] = `
<div
  className="Tabs___horizontal___lg"
/>
`;
